integration-tests/load/ccip/helpers.go

```diff
@@ -0,0 +1,315 @@
+package ccip
+
+import (
+	"context"
+	"errors"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	chainselectors "github.com/smartcontractkit/chain-selectors"
+
+	"math/big"
+
+	"github.com/smartcontractkit/chainlink-ccip/pkg/types/ccipocr3"
+	"github.com/smartcontractkit/chainlink-common/pkg/logger"
+	"github.com/smartcontractkit/chainlink-testing-framework/wasp"
+	"github.com/smartcontractkit/chainlink/deployment"
+	"github.com/smartcontractkit/chainlink/v2/core/gethwrappers/ccip/generated/offramp"
+)
+
+var wgroup sync.WaitGroup
+
+const (
+	transmitted = iota
+	committed
+	executed
+	LokiLoadLabel  = "ccip_load_test"
+	ErrLokiClient  = "failed to create Loki client for monitoring"
+	ErrLokiPush    = "failed to push metrics to Loki"
+	tickerDuration = 30 * time.Second
+)
+
+// todo: Have a different struct for commit/exec?
+type LokiMetric struct {
+	EventType      int       `json:"event_type"`
+	Timestamp      time.Time `json:"timestamp"`
+	GasUsed        uint64    `json:"gas_used"`
+	SequenceNumber uint64    `json:"sequence_number"`
+}
+
+func SendMetricsToLoki(l logger.Logger, lc *wasp.LokiClient, updatedLabels map[string]string, metrics *LokiMetric) {
+	if err := lc.HandleStruct(wasp.LabelsMapToModel(updatedLabels), time.Now(), metrics); err != nil {
+		l.Error(ErrLokiPush)
+	}
+}
+
+func setLokiLabels(src, dst uint64) (map[string]string, error) {
+	srcChainID, err := chainselectors.GetChainIDFromSelector(src)
+	if err != nil {
+		return nil, err
+	}
+	dstChainID, err := chainselectors.GetChainIDFromSelector(dst)
+	if err != nil {
+		return nil, err
+	}
+	return map[string]string{
+		"sourceEvmChainId":    srcChainID,
+		"sourceSelector":      strconv.FormatUint(src, 10),
+		"destEvmChainId":      dstChainID,
+		"destinationSelector": strconv.FormatUint(dst, 10),
+		"testType":            LokiLoadLabel,
+	}, nil
+}
+
+type finalSeqNrReport struct {
+	sourceChainSelector uint64
+	expectedSeqNrRange  ccipocr3.SeqNumRange
+}
+
+func subscribeDeferredCommitEvents(
+	ctx context.Context,
+	lggr logger.Logger,
+	offRamp *offramp.OffRamp,
+	srcChains []uint64,
+	startBlock *uint64,
+	loki *wasp.LokiClient,
+	chainSelector uint64,
+	client deployment.OnchainClient,
+	finalSeqNrs chan finalSeqNrReport,
+	errChan chan error,
+) {
+	defer wgroup.Done()
+
+	lggr.Infow("starting commit event subscriber for ",
+		"destChainSelector", chainSelector,
+		"offRamp", offRamp.Address().String(),
+		"startblock", startBlock,
+	)
+	seenMessages := make(map[uint64][]uint64)
+	expectedRange := make(map[uint64]ccipocr3.SeqNumRange)
+	completedSrcChains := make(map[uint64]bool)
+	for _, srcChain := range srcChains {
+		// todo: seenMessages should hold a range to avoid hitting memory constraints
+		seenMessages[srcChain] = make([]uint64, 0)
+		completedSrcChains[srcChain] = false
+	}
+
+	sink := make(chan *offramp.OffRampCommitReportAccepted)
+	subscription, err := offRamp.WatchCommitReportAccepted(&bind.WatchOpts{
```

Comment:
nit : have you checked  `event.Resubscribe`? it comes handy when the subscription gets into flakey rpc error for longer duration run